sdks/java/io/hcatalog/src/main/java/org/apache/beam/sdk/io/hcatalog/HCatalogIO.java

```diff
@@ -0,0 +1,507 @@
+     */
+    public Read withConfigProperties(Map<String, String> configProperties) {
+      return toBuilder().setConfigProperties(Collections.unmodifiableMap(configProperties))
+          .build();
+    }
+
+    /**
+     * Sets the database name.
+     * This is optional, assumes 'default' database if none specified
+     * @param database
+     * @return
+     */
+    public Read withDatabase(String database) {
+      return toBuilder().setDatabase(database).build();
+    }
+
+    /**
+     * Sets the table name to read from.
+     * This is mandatory
+     * @param table
+     * @return
+     */
+    public Read withTable(String table) {
+      return toBuilder().setTable(table).build();
+    }
+
+    /**
+     * Sets the filter (partition) details.
+     * This is optional, assumes none if not specified
+     * @param filter
+     * @return
+     */
+    public Read withFilter(String filter) {
+      return toBuilder().setFilter(filter).build();
+    }
+
+    Read withSplitId(int splitId) {
+      checkArgument(splitId >= 0, "Invalid split id-" + splitId);
+      return toBuilder().setSplitId(splitId).build();
+    }
+
+    Read withContext(ReaderContext context) {
+      return toBuilder().setContext(context).build();
+    }
+
+    @Override
+    public PCollection<DefaultHCatRecord> expand(PBegin input) {
+      return input.apply(org.apache.beam.sdk.io.Read.from(new BoundedHCatalogSource(this)));
+    }
+
+    @Override
+    public void validate(PipelineOptions options) {
+      checkNotNull(getTable(), "table");
+      checkNotNull(getConfigProperties(), "configProperties");
+    }
+
+    @Override
+    public void populateDisplayData(DisplayData.Builder builder) {
+      super.populateDisplayData(builder);
+      builder.add(DisplayData.item("configProperties", getConfigProperties().toString()));
+      builder.add(DisplayData.item("table", getTable()));
+      builder.addIfNotNull(DisplayData.item("database", getDatabase()));
+      builder.addIfNotNull(DisplayData.item("filter", getFilter()));
+    }
+  }
+
+  /**
+   * A HCatalog {@link BoundedSource} reading {@link DefaultHCatRecord} from  a given instance.
+   */
+  @VisibleForTesting
+  static class BoundedHCatalogSource extends BoundedSource<DefaultHCatRecord> {
+    private Read spec;
+
+    BoundedHCatalogSource(Read spec) {
+      this.spec = spec;
+    }
+
+    @Override
+    public Coder<DefaultHCatRecord> getDefaultOutputCoder() {
+      return WritableCoder.of(DefaultHCatRecord.class);
+    }
+
+    @Override
+    public void validate() {
+      spec.validate(null);
+    }
+
+    @Override
+    public void populateDisplayData(DisplayData.Builder builder) {
+      spec.populateDisplayData(builder);
+    }
+
+    @Override
+    public BoundedReader<DefaultHCatRecord> createReader(PipelineOptions options) {
+      return new BoundedHCatalogReader(this);
+    }
+
+    /**
+     * This implementation simply returns 0L as this is not used in calculating the splits.
```

Comment:
If this `BoundedSource` is unable to dynamically split, unable to estimate size, and unable to use `desiredBundleSizeBytes` when splitting, then it shouldn't be a `BoundedSource`, it should simply be a composition of two `DoFn`'s (with a shuffle in the middle to ensure parallelism) - see e.g. how `DatastoreIO` is implemented: there's one `DoFn` to split and another one to read.

`BoundedSource` is an advanced and error-prone API and should be used only when strictly necessary. See https://beam.apache.org/documentation/io/authoring-overview/#when-to-implement-using-the-source-api